javaagent-bootstrap/src/main/java/io/opentelemetry/javaagent/OpenTelemetryAgent.java

```diff
@@ -64,24 +65,30 @@ private static void startAgent(Instrumentation inst, boolean fromPremain) {
 
   private static synchronized File installBootstrapJar(Instrumentation inst)
       throws IOException, URISyntaxException {
-
-    CodeSource codeSource = OpenTelemetryAgent.class.getProtectionDomain().getCodeSource();
-
-    if (codeSource == null) {
-      throw new IllegalStateException("could not get agent jar location");
+    // we are not using OpenTelemetryAgent.class.getProtectionDomain().getCodeSource() to get agent
+    // location because getProtectionDomain does a permission check with security manager
+    URL url =
+        OpenTelemetryAgent.class
+            .getClassLoader()
+            .getResource(OpenTelemetryAgent.class.getName().replace('.', '/') + ".class");
+    if (url == null || !"jar".equals(url.getProtocol())) {
+      throw new IllegalStateException("could not get agent jar location from url " + url);
     }
-
-    File javaagentFile = new File(codeSource.getLocation().toURI());
+    String resourcePath = url.toURI().getSchemeSpecificPart();
+    int protocolSeparatorIndex = resourcePath.indexOf(":");
+    int resourceSeparatorIndex = resourcePath.indexOf("!/");
+    if (protocolSeparatorIndex == -1 || resourceSeparatorIndex == -1) {
+      throw new IllegalStateException("could not get agent location from url " + url);
+    }
+    String agentPath = resourcePath.substring(protocolSeparatorIndex + 1, resourceSeparatorIndex);
+    File javaagentFile = new File(agentPath);
 
     if (!javaagentFile.isFile()) {
       throw new IllegalStateException(
           "agent jar location doesn't appear to be a file: " + javaagentFile.getAbsolutePath());
     }
 
-    // passing verify false for vendors who sign the agent jar, because jar file signature
-    // verification is very slow before the JIT compiler starts up, which on Java 8 is not until
-    // after premain execution completes
-    JarFile agentJar = new JarFile(javaagentFile, false);
+    JarFile agentJar = ((JarURLConnection) url.openConnection()).getJarFile();
```

Comment:
does this now reuse the underlying `JarFile` that was already opened for reading `OpenTelemetryAgent` class file? and by doing so bypass (re-)verification?

Reply:
I thought this would have access to an already open `JarFile`, but debugger shows that it is opening a new one. Reverted this back to `new JarFile(javaagentFile, false);`
There are more places where jar files are open with verification. When the main agent class (`OpenTelemetryAgent`) is loaded then the agent jar is open with verification. I think this could be avoided by adding the agent jar to boot classpath, could use `Boot-Class-Path` manifest attribute. When byte buddy starts retransforming then class file locators call `getResourceAsStream` on `BootstrapClassLoaderProxy` which results in jar verification (adding agent to boot loader should also fix these).
If you are serious about speeding up jdk8 startup with jar verification then your best bet is to enable the delay agent start thing in `AgentInitializer` for all jdk8.

Reply:
> If you are serious about speeding up jdk8 startup with jar verification then your best bet is to enable the delay agent start thing in `AgentInitializer` for all jdk8.

oh, good idea, I'll try this out, thx